README.rst
htof
===============

This repo contains htof, the package for parsing intermediate data from the Gaia and
Hipparcos satellites, and reproducing five, seven, and nine (or higher) parameter fits to their astrometry.

.. image:: https://coveralls.io/repos/github/gmbrandt/HTOF/badge.svg?branch=main
    :target: https://coveralls.io/github/gmbrandt/HTOF?branch=main


.. image:: https://app.travis-ci.com/gmbrandt/HTOF.svg?branch=main
    :target: https://app.travis-ci.com/gmbrandt/HTOF

Parallax is handled by the :code:`sky_path` module which was written by Anthony Brown
as a part of his astrometric-sky-path package: https://github.com/agabrown/astrometric-sky-path/

Installation
------------
htof can be installed with PyPi and pip, by running

.. code-block:: bash

    pip install htof

or by running

.. code-block:: bash

    pip install .


while in the root directory of this repo. It can also be installed directly from github using

.. code-block:: bash

    pip install git+https://github.com/gmbrandt/htof

Usage
-----

HTOF has a rich variety of usages. We encourage the reader to consult the examples.ipynb jupyter notebook
for a set of usage examples (e.g., fitting the standard astrometric model to data, combining astrometric missions).
However, we also go into a few basic and specific use cases in this readme. Also see
examples_recalibrating_hip2 and GenerateSyntheticGaiaAstrometry for more uses of htof.

If you use HTOF, please cite the zenodo reference (https://doi.org/10.5281/zenodo.4104383) and
the source paper (https://ui.adsabs.harvard.edu/abs/2021AJ....162..230B/abstract)

Usage: Fits without Parallax
----------------------------
The following examples show how one would both load in and fit a line to the intermediate astrometric data (IAD)
from either Hipparcos data reduction or Gaia.

HTOF will download any missing IAD files for you (for Hip1, Hip2 (Java tool data), and Gaia). You should
provide a valid directory though for htof to save the file into for future use. Currently, for the
automatic download to work, you must provide a hipparcos name for the source (e.g., 27321).
Include ONLY the numeric part of the name. For an example of downloading Gaia data, see Example 6 in examples.ipynb

**For Gaia**:

If the automatic download of the GOST scanning law does not work, or the source does not have a
hipparcos ID. then you will have to download the GOST file manually with the user interface. Download a .csv of the
predicted scans and scan epochs from GOST (https://gaia.esac.esa.int/gost/). In particular, using the 'submit for
events forecast' feature on the website. One should select the widest range of dates
possible because htof automatically restricts the predicted epochs of observations
to the desired data release range (e.g., EDR3) and removes any astrometric gaps.

Let ra_vs_epoch, dec_vs_epoch be 1d arrays of ra and dec positions. Assume we want to fit to data from GaiaDR2 on the
star with hip id 027321. The following lines parse the intermediate data and fit a line.

.. code-block:: python

    from htof.main import Astrometry
    import numpy as np
    astro = Astrometry('GaiaDR2', '027321', 'htof/test/data_for_tests/GaiaDR2/IntermediateData', format='jyear')  # parse
    # note that if you do not have a GOST csv file with 027321 in the name, inside of
    # 'htof/test/data_for_tests/GaiaDR2/IntermediateData' , then htof will download it for you automatically!
    ra_vs_epoch = dec_vs_epoch = np.zeros(len(astro.data), dtype=float) # dummy set of ra and dec to fit.
    ra0, dec0, mu_ra, mu_dec = astro.fit(ra_vs_epoch, dec_vs_epoch)


:code:`GaiaeDR3` will select all data corresponding to the eDR3 data interval and exclude
eDR3 deadtimes. :code:`GaiaDR2` will select all data corresponding to the DR2 data interval (excluding dead times).
Finally, :code:`Gaia` will select all the data present in the GOST predicted observation file.

Note, selecting :code:`Gaia` and allowing HTOF to download the GOST scanning law may result in a wierd coverage of the
scanning law. We recommend sticking with defined data release intervals, so :code:`GaiaeDR3`, etc.

In the :code:`astro.fit` part, ra_vs_epoch and dec_vs_epoch are the positions in right ascension and declination of the object.
These arrays must have the same shape as astro.data.julian_day_epoch(),
which are the epochs in the intermediate data. :code:`format='jyear'` specifies
the time units of the output best fit parameters. The possible choices of format
are the same as the choices for format in astropy.time.Time(val, format=format). If :code:`format='jyear'`, then the output :code:`mu_ra`
would have units of mas/year. If :code:`jd` then the output is mas/day. Both Hipparcos and Gaia catalogs list parallaxes
in milli-arcseconds (mas), and so positional units are always in mas for HTOF.

**For Hipparcos** :

:code:`Hip2` refers to the DVD IAD which is now obsolete. :code:`Hip21` refers to the
Java Tool Intermediate Astrometric Data (IAD) and best fit parameters. This is the preferred set of
data to use with the 2007 re-reduction (preferred over the DVD IAD). The Hipparcos Java Tool data parser is meant for
the 2014 Java tool data (Java tool first released at
https://www.cosmos.esa.int/web/hipparcos/java-tools/intermediate-data, in 2014). As of 2021, there has not been an
update to the Java tool.

Remember, Hip21 will be automatically downloaded
per source. But if you want to avoid downloading the data on the fly, then the full Java Tool Intermediate Astrometric Data can be downloaded from
https://www.cosmos.esa.int/web/hipparcos/hipparcos-2 and extracted (ignore the _MACOSX folder if there is one).
One would then point any HTOF parser to the ResRec_JavaTool folder that contains the H00 etc. subfolders of the individual IAD files. So:

.. code-block:: python

    from htof.main import Astrometry
    astro = Astrometry('Hip21', star_id='027321', '/home/user/Downloads/ResRec_JavaTool_2014/ResRec_JavaTool_2014', format='jyear')  # parse
    ra0, dec0, mu_ra, mu_dec = astro.fit(ra_vs_epoch, dec_vs_epoch)



We discuss enabling fits with parallax later. By default, the fit is a four-parameter fit: it returns the parameters to the line of best
fit to the sky path ra_vs_epoch, dec_vs_epoch. If you want a 6 parameter or 8 parameter fit, specify
fit_degree = 2 or fit_degree = 3 respectively. E.g.

.. code-block:: python

    from htof.main import Astrometry
    astro = Astrometry('GaiaDR2', '027321', 'htof/test/data_for_tests/GaiaDR2/IntermediateData', format='jyear',
                       fit_degree=2)
    ra0, dec0, mu_ra, mu_dec, acc_ra, acc_dec = astro.fit(ra_vs_epoch, dec_vs_epoch)

If fit_degree = 3, then the additional last two parameters would be the jerk in right ascension and declination, respectively.
The sky path in RA (for instance) should be reconstructed by `ra0 + mu_ra*t + 1/2*acc_ra*t**2` where `t` are the epochs
from `astro.fitter.epoch_times` minus the central epoch for RA (if provided).

HTOF allows fits of arbitrarily high degree. E.g. setting fit_degree=3 would give a 9 parameter
fit (if using parallax as well). One should specify a central epoch for the fit, typically choosing the central epoch
from the catalog (e.g. 2015.5 for GaiaDR2, 2016 for GaiaEDR3, 1991.25 for Hipparcos). You can specify the central epoch by:

.. code-block:: python

    from htof.main import Astrometry

    astro = Astrometry('GaiaDR2', '027321', 'htof/test/data_for_tests/GaiaDR2/IntermediateData',
                       central_epoch_ra=2015.5, central_epoch_dec=2015.5, format='jyear')
    ra0, dec0, mu_ra, mu_dec = astro.fit(ra_vs_epoch, dec_vs_epoch)

The format of the central epochs must be specified along with the central epochs. The best fit sky path in right ascension would then be
:code:`ra0 + mu_ra * (epochs - centra_epoch_ra)`. The central epoch matters for numerical stability and covariances.
E.g., dont choose a central epoch like the year 1200 for GaiaDR2. One should almost always choose the central epoch
from the catalog.

One can access the BJD epochs with

.. code-block:: python

    astro.data.julian_day_epoch()

If you want the standard (1-sigma) errors on the parameters, set :code:`return_all=True` when fitting:

.. code-block:: python

    from htof.main import Astrometry

    astro = Astrometry('GaiaDR2', '027321', 'htof/test/data_for_tests/GaiaDR2/IntermediateData',
                        central_epoch_ra=2015.5, central_epoch_dec=2015.5, format='jyear')
    solution_vector, errors, chisq, residuals = astro.fit(ra_vs_epoch, dec_vs_epoch, return_all=True)


`errors` is an array the same shape as solution_vector, where each entry is the 1-sigma error for the
parameter at the same location in the solution_vector array. `chisq` is the formal chisquared of the
fit to the data, and `residuals` are the data - model residuals (given as a Nx2 shaped array, where N is the number
of transits. The first column are the ra residuals and the second are the declination residuals).
For a simple refit to the catalog IAD, `residuals` (converted to the AL basis) will equal (up to round off)
the residuals given in the IAD.  One could convert the residuals to the along scan basis by doing:

.. code-block:: python

    from htof.special_parse import to_along_scan_basis
    residuals = to_along_scan_basis(ra_decresiduals[:, 0], ra_decresiduals[:, 1], astro.data.scan_angle.values)
    # now residuals will be a one dimensional array of length N (number of transits), giving the residuals along the
    # scan.

For Hip1 and Hip21, HTOF loads in the real
catalog errors and so the parameter error estimates (`errors`) should match those given in the catalog. For Hip2, the
along scan errors are automatically inflated or deflated in accordance with D. Michalik et al. 2014.
For Gaia we do not have the error estimates from the GOST tool. The AL errors are set to 1 mas by default and so the
best-fit parameter errors to Gaia will not match those reported by the catalog.


`chisq` is the chi-squared of the fit (the sum of `(data - model)^2/error^2`). The `chisq` from `astro.fit`
should equal (for Hip1 and Hip2) the chi-squared calculated from the intermediate data:

.. code-block:: python

    chisq = np.sum(astro.data.residuals ** 2 / astro.data.along_scan_errs ** 2)

Saving processed intermediate data
----------------------------------
To save the scan angles, residuals, along-scan errors, inverse covariance matrices, and julian day
epochs, one can call ``Astrometry.data.write(path)`` to write out the data, where path is a string which
points to the full filepath including the data extension. We recommend ``.csv``, however any file extension
supported by ``astropy.table.Table.write()`` is supported. As well, one can call ``Astrometry.data.write(path)``
with any of the kwargs or args of ``astropy.table.Table.write()``.

Usage: Fits with Parallax
-------------------------
To fit an object with parallax, there are two ways. Both are equivalent.

    1. Let htof compute the parallax factors anew.
    2. Pull the parallax factors from the IAD or the scanning law.

Option 1: we need to provide a `central_ra` and `central_dec` to the `Astrometry` class. These positions
will be used to calculate the parallax components of the fit (the parallax factors). Using beta pic as an example,
we would do:


.. code-block:: python

    from htof.main import Astrometry
    import numpy as np
    from astropy.coordinates import Angle
    # central ra and dec from the Hip1 catalog
    cntr_ra, cntr_dec = Angle(86.82118054, 'degree'), Angle(-51.06671341, 'degree')
    # generate fitter and parse intermediate data
    astro = Astrometry('Hip1', '27321', 'htof/test/data_for_tests/Hip1/IntermediateData', central_epoch_ra=1991.25,
                       central_epoch_dec=1991.25, format='jyear', fit_degree=1, use_parallax=True,
                       central_ra=cntr_ra, central_dec=cntr_dec)
    ra_vs_epoch = dec_vs_epoch = np.zeros(len(astro.data), dtype=float) # dummy set of ra and dec to fit.
    solution_vector, errors, chisq, residuals = astro.fit(ra_vs_epoch, dec_vs_epoch, return_all=True)
    parallax, ra0, dec0, mu_ra, mu_dec = solution_vector


Option 2: In most use cases, this option is perfectly fine. And it is simpler. The object's parallax factors is available with the IAD (or the scanning law in the case of Gaia)
So you do not need a `central_ra` and `central_dec` to the `Astrometry` class. In which case, you can do:

.. code-block:: python

    from htof.main import Astrometry
    import numpy as np
    # generate fitter and parse intermediate data
    astro = Astrometry('Hip1', '27321', 'htof/test/data_for_tests/Hip1/IntermediateData', central_epoch_ra=1991.25,
                       central_epoch_dec=1991.25, format='jyear', fit_degree=1, use_parallax=True,
                       use_catalog_parallax_factors=True)
    ra_vs_epoch = dec_vs_epoch = np.zeros(len(astro.data), dtype=float) # dummy set of ra and dec to fit.
    solution_vector, errors, chisq, residuals = astro.fit(ra_vs_epoch, dec_vs_epoch, return_all=True)
    parallax, ra0, dec0, mu_ra, mu_dec = solution_vector

Note that we have set ``use_catalog_parallax_factors=True``. This tells htof to *not* compute parallax factors
anew, and instead to pull them from the IAD. If data choice was 'Gaiaedr3' instead of 'hip1', then the parallax factors
would come from the GOST csv file.

Appendix
--------

Parsing and fitting manually
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The Astrometry object is essentially just a wrapper for data parsing and fitting all in one, and consequently
could be limiting. This section describes how to reproduce Astrometry.fit by accessing the data parser objects and
the fitter object separately. You would do this if, for instance, you did not want to use
the built-in parallax motions generated by HTOF. Or if, you wanted to do a GaiaEDR3 fit with your own AL errors.
I show here how to reproduce a five-parameter fit.


.. code-block:: python

    from htof.parse import HipparcosOriginalData # or GaiaData or HipparcosReReduction
    data = HipparcosOriginalData()
    data.parse(star_id='004391', intermediate_data_directory='htof/test/data_for_tests/Hip1/IntermediateData/')
    data.calculate_inverse_covariance_matrices()

data now has a variety of intermediate data products such as the scan angles, the epochs when each
data point was collected, the inverse covariance matrices describing the errors of the scan,
and the BJD epochs accessible through :code:`data.julian_day_epoch()`.

You could modify the along-scan errors (let's say if you were doing a Gaia DR4/DR5 forecast) with:

.. code-block:: python


CHANGES.md
1.1.5 (2022-06-24)
------------------
- Added the ability to download the Hipparcos re-reduction data (Java Tool Version). 
- Slight refactor of the internals of downloading Gaia/Hip1 files from the web.
- Crucial bug fix in get_intermediate_data_filename that sometimes would give the wrong IAD file
for small Hip ids.
- ONLY numeric star_ids are allowed in Parsers. E.g., 2732 is allowed and not H88iJJ2.

1.1.4 (2022-06-03)
------------------
- Misc bug fixes for data downloading, and warning fixes.

1.1.3 (2022-05-25)
------------------
- HipparcosOriginalData Parser classes can now automatically download the Hip 1997 IAD.

1.1.2 (2022-05-23)
------------------
- Updated the recalibration parameters.

1.1.1 (2022-05-10)
------------------
- Gaia Parser classes now can fetch GOST scanning law from the GOST API. 
- Downloaded GOST scanning laws are saved to the directory provided.
- having the examples in an examples/ folder broke the filepaths.

1.1.0 (2022-04-19)
------------------
- Added a new hip2 parser, in special_parse.py, that will add the residual offset and cosmic dispersion of 
  Brandt et al. 2022 and output new IAD with new best-fit parameters. Works for 5, 7, and 9 parameter sources.
- Added a new hip2 parser that autodetects whether the data file is the java tool or DVD. Technically, it is 
  a parser factory.
- Added the option for htof.main.Astrometry to load in and use the parallax factors supplied with
the hipparcos IAD.
- htof.special_parse.Hipparcos2Recalibrated can now write out a data file in the same format 
  (although they do not match byte-by-byte exactly) as the hip2 java tool IAD.
- fitting with return_all=True now returns 4 arrays, the last is the residuals (in ra and dec) of the fit.
- New examples jupyter notebook for recalibrating the hip2 java tool IAD.
- Astrometry() now has an along_scan_error_scaling attribute that will scale the
  along scan errors by that factor. This is particularly useful for gaia forecasting. Setting
  along_scan_error_scaling = 0.1 for instance, will artifically make all the Gaia GOST 
  scans have an error of 0.1 mas (normally, these along scan errors are not provided by GOST)

1.0.3 (2022-01-26)
------------------
- updated the jupyter examples notebook. Included a central epoch comparison to the Gaia archive.

1.0.2 (2021-12-20)
------------------
- Added Python 3.9 to travis testing suite. As of writing, python 3.10 testing is
not available with travis. I manually tested that pytest-sv works with 
  python 3.10.1 on my local ubuntu 20.04 machine.

1.0.1 (2021-09-01)
------------------
- updated docs.
- Added gaiaEDR3 ephemeris, resolving a bug when using use_parallax=True and gaiaEDR3.

1.0.0 (2021-09-01)
------------------
- Updated the parsing method to read in the new, final version of the Hip2 Java tool data that
will be published on the ESA website.
- Changed the Hip2 (DVD) refit so that the parameter errors are saved, not the parameter differences.
- Changed back the error inflation computation to use the catalog f2 value.

0.4.2 (2021-06-24)
------------------
- Introduced a .meta attribute to data parsers. This holds things like the catalog f2 value.
- Updated examples.ipynb.

0.4.1 (2021-06-15)
------------------
- Improved warning messages for the adhoc correction.

0.4.0 (2021-04-30)
------------------
- Implemented a better data fix for Hip2 java data. This fixes most of the 6400 discrepant sources
that have the Hip2 file-write error. Note that this is not perfect, there are some minor
  degeneracies between which epochs to reject. See the note in parse.find_epochs_to_reject()
- Note that this write-out bug fix does not work on the dvd data (note, 
  the dvd still has the same bug, it is just not easily correctable).
- Updated the flagged source list (hip2_javatool_flagged.txt).
- Parallax factors from the IAD are now loaded on data.parse()
- Updated parse to use the new version of the java tool IAD. Old versions of the java
tool IAD will not work.

0.3.5 (2021-04-19)
------------------
- Fixed improper syntax calls to panda etc. so that warnings are silence in python 3.8 and beyond.
- We have a deprecated call that cannot be fixed, and appears to be
  located upstream in ipython https://github.com/ipython/ipykernel/issues/560
- Removed normed option for fitting.

0.3.4 (2021-02-24)
------------------
- Removed the numba speed up from FastFitter. Will add it back in if there is ever a need for a fast fitter.

0.3.3 (2020-12-23)
------------------
- Added Gaia eDR3 deadtimes and a GaiaeDR3 class for Gaia early data release3.
- Added a 'gaiaedr3' data choice to the main.Astrometry class.
- Added the method AstrometricFitter.find_optimal_central_epoch() which will find
the central epoch that minimizes the covariance between position and proper motion
for either right-ascension (if calling find_optimal_central_epoch('ra')) or declination 
(if calling find_optimal_central_epoch('dec')).
- Added a convienience function to Astrometry: Astrometry.optimal_central_epochs() that
returns a dictionary with the epochs in ra and dec that give minimum (close to 0, like 1e-10)
covariance between ra and mu_ra, and dec and mu_dec respectively.

0.3.2 (2020-10-22)
-------------------
- Added a 'BOTH' data_choice option to the HipparcosOriginalData DataParser class. Selecting
'BOTH' will keep both NDAC and FAST data, i.e. it will leave the iad unmerged.
- Fixed typo in the flagged hip2 source list filename.

0.3.1 (2020-10-22)
-------------------
- Fixed a typo in the path to the package data for the dead time table.

0.3.0 (2020-10-17)
-------------------
- HipparcosRereductionCDBook has been renamed HipparcosRereductionDVDBook because the data come from a DVD.
- Added various utilities and scripts for validating HTOF's fitting and parsing 
routines over the entire Hipparcos catalog.
- Added a list of flagged sources for which HTOF cannot reproduce their catalog solution.

htof/special_parse.py
import numpy as np
import warnings
import pkg_resources
from pandas import DataFrame, Series
from shutil import copy
from copy import deepcopy

from astropy.time import Time
from astropy.coordinates import Angle
from astropy import units as u
from scipy import stats, special
from astropy.table import Table
from htof.fit import AstrometricFitter
from htof.parse import HipparcosRereductionJavaTool, HipparcosRereductionDVDBook, DataParser


class Hipparcos2Recalibrated(HipparcosRereductionJavaTool):
    """
    A parser class which re-calibrates the hipparcos 2 data according to Brandt et al. 2022

    This is a monstrous parser class because well it is doing a lot.

    Note that this "Parser" class actually refits the data. This is programmatically poorly set up,
    in the sense that we parse the data and fit it, mixing this class with an AstrometricFitter class... For now,
    this is OK because the objective of this class is mainly so other users can save and inspect the recalibrated
    hipparcos 2 IAD.

    """
    EPOCHREJECTLIST = Table.read(pkg_resources.resource_filename('htof',
                                                                 'data/epoch_reject_shortlist.csv'), format='ascii')

    def __init__(self, scan_angle=None, epoch=None, residuals=None, inverse_covariance_matrix=None,
                 along_scan_errs=None, meta=None, residual_offset=0.141, cosmic_dispersion=2.25):
        super(Hipparcos2Recalibrated, self).__init__(scan_angle=scan_angle, along_scan_errs=along_scan_errs,
                                                     epoch=epoch, residuals=residuals,
                                                     inverse_covariance_matrix=inverse_covariance_matrix,
                                                     meta=meta)
        self.residual_offset = residual_offset
        self.cosmic_dispersion = cosmic_dispersion
        self.recalibrated_header = None
        self.recalibrated_data = None

    def parse(self, star_id, intermediate_data_directory, attempt_adhoc_rejection=True,
              reject_known=True, **kwargs):
        # important that the 2007 error inflation is turned off (error_inflate=False)
        header, raw_data = super(Hipparcos2Recalibrated, self).parse(star_id, intermediate_data_directory,
                                                                     error_inflate=False,
                                                                     attempt_adhoc_rejection=attempt_adhoc_rejection,
                                                                     reject_known=reject_known, **kwargs)
        apply_calibrations = True
        if not (self.meta['catalog_soltype'] == 5 or self.meta['catalog_soltype'] == 7 or self.meta['catalog_soltype'] == 9):
            warnings.warn(f'This source has a solution type of {self.meta["catalog_soltype"]}. '
                          f'htof will only recalibrate 5, 7, and 9 parameter solutions currently. '
                          f'No recalibration will be performed.')
            apply_calibrations = False
        if attempt_adhoc_rejection is False or reject_known is False:
            warnings.warn('We have not tested recalibration without rejecting any of the flagged or bugged '
                          'observations. No recalibration will be performed.')
            apply_calibrations = False

        if apply_calibrations:
            # apply the calibrations
            self.residuals += self.residual_offset  # note that this modifies the raw_data column also.
            self.along_scan_errs = np.sqrt(self.along_scan_errs**2 + self.cosmic_dispersion**2)
            self.calculate_inverse_covariance_matrices()
            # munge the parallax factors into the correct form. Note that we are using
            # the parallax factors from the catalog here to keep everything consistent.
            ra_motion, dec_motion = to_ra_dec_basis(self.parallax_factors.values, self.scan_angle.values)
            parallactic_perturbations = {'ra_plx': ra_motion, 'dec_plx': dec_motion}
            # refit the data, calculate the new residuals, and parameters.
            fit_degree = {5: 1, 7: 2, 9: 3}[int(self.meta['catalog_soltype'])]
            fitter = AstrometricFitter(inverse_covariance_matrices=self.inverse_covariance_matrix,
                                       epoch_times=Time(Time(self.julian_day_epoch(), format='jd'), format='jyear').value,
                                       central_epoch_dec=1991.25,
                                       central_epoch_ra=1991.25,
                                       fit_degree=fit_degree,
                                       use_parallax=True,
                                       parallactic_pertubations=parallactic_perturbations)
            # get residuals in ra and dec.
            ra = Angle(self.residuals.values * np.sin(self.scan_angle.values), unit='mas')
            dec = Angle(self.residuals.values * np.cos(self.scan_angle.values), unit='mas')
            # fit the residuals
            coeffs, errors, Q, new_residuals = fitter.fit_line(ra.mas, dec.mas, return_all=True)
            # compute the along-scan residuals
            new_residuals = to_along_scan_basis(new_residuals[:, 0], new_residuals[:, 1], self.scan_angle.values)
            self.residuals = Series(new_residuals, index=self.residuals.index)
            """
            update the header with the new statistics
            """
            ntransits, nparam = len(self), int(self.meta['catalog_soltype'])
            header['second']['NOB'] = len(self)
            header['second']['NR'] = 0  # because we automatically remove any "flagged as rejected" observations.
            header['first']['F1'] = 0
            header['first']['NRES'] = len(self)
            header['first']['F2'] = special.erfcinv(stats.chi2.sf(Q, ntransits - nparam)*2)*np.sqrt(2)
            if np.isfinite(header['first']['F2']):
                header['first']['F2'] = np.round(header['first']['F2'], 4)
            # update the best fit parameters with the new values
            # dpmRA  dpmDE  e_dpmRA  e_dpmDE  ddpmRA  ddpmDE  e_ddpmRA  e_ddpmDE
            for i, key, dp, in zip(np.arange(nparam),
                                   ['Plx', 'RAdeg', 'DEdeg', 'pm_RA', 'pm_DE', 'dpmRA', 'dpmDE', 'ddpmRA', 'ddpmDE'],
                                   [2, 8, 8, 2, 2, 2, 2, 2, 2]):
                if key != 'RAdeg' and key != 'DEdeg':
                    header['third'][key] = np.round(header['third'][key] + coeffs[i], dp)
                else:
                    # convert the perturbation from mas to degrees.
                    header['third'][key] = np.round(header['third'][key] + (coeffs[i] * u.mas).to(u.degree).value, dp)
            # update the errors with the new errors
            for i, key, dp, in zip(np.arange(nparam),
                                   ['e_Plx', 'e_RA', 'e_DE', 'e_pmRA', 'e_pmDE', 'e_dpmRA', 'e_dpmDE', 'e_ddpmRA', 'e_ddpmDE'],
                                   [2, 2, 2, 2, 2, 2, 2, 2, 2]):
                header['third'][key] = np.round(errors[i], dp)
            # save the modified header to the class, because these will be
            # used by self.write_as_javatool_format()
            self.recalibrated_header = deepcopy(header)
            """
            update the raw_data columns with the new data. Note that rejected/bugged epochs are already
            taken care of.
            """
            # data order in Java tool data: IORB   EPOCH    PARF    CPSI    SPSI     RES   SRES
            recalibrated_data = DataFrame({'1': self._iorb, '2': self._epoch - 1991.25, '3': self.parallax_factors,
                                           '4': self._cpsi, '5': self._spsi, '6': np.round(self.residuals.values, 3),
                                           '7': np.round(self.along_scan_errs.values, 3)})
            self.recalibrated_data = recalibrated_data
            header, raw_data = None, None  # the raw header and raw data have been modified, so clear them.
        return header, raw_data

    def write_as_javatool_format(self, path: str):
        """
        Variant of the .write() method that is specially for this Parser. It writes out the fixed data
        with the exact same format as the input JavaTool IAD (i.e., the .d files).
        We recommend using the normal .write() format. This method is for users who do not want to adopt the
        htof file format output by the normal .write() format.

        :param: path: path to write the file to. filename should end with .d . ".d" is the same file ending as the
        java tool IAD. But in principle, most common ascii file extensions should work (e.g., ".txt" etc).
        """
        if self.recalibrated_header is None or self.recalibrated_data is None:
            warnings.warn('This source was NOT recalibrated, see earlier warnings as to why. Will not save any '
                          f'output file at {path} because no recalibration was done.')
            return None
        # order of the header values:
        ks1 = ['HIP', 'MCE', 'NRES', 'NC', 'isol_n', 'SCE', 'F2', 'F1']
        ks2 = ['Hp', 'B-V', 'VarAnn', 'NOB', 'NR']
        ks3 = ['RAdeg', 'DEdeg', 'Plx', 'pm_RA', 'pm_DE', 'e_RA', 'e_DE', 'e_Plx', 'e_pmRA', 'e_pmDE',
               'dpmRA', 'dpmDE', 'e_dpmRA', 'e_dpmDE', 'ddpmRA', 'ddpmDE', 'e_ddpmRA', 'e_ddpmDE',
               'upsRA', 'upsDE', 'e_upsRA', 'e_upsDE', 'var']
        # byte entries of each header value. Stored as (start, stop) inclusive intervals.
        # the exact bytes are stored, according to the IAD readme, so that the output format is exactly the same.
        bs1 = [(3, 8), (10, 15), (17, 19), (22, 22), (25, 27), (32, 35), (37, 42), (44, 45)]
        bs2 = [(3, 9), (11, 16), (18, 18), (25, 27), (29, 31)]
        bs3 = [(3, 14), (16, 27), (29, 36), (38, 45), (47, 54),
               (56, 61), (63, 68), (70, 75), (77, 82), (84, 89), (91, 96),
               (98, 103), (105, 110), (114, 119), (123, 128), (131, 136), (139, 144),
               (149, 154), (159, 164), (167, 172), (175, 180), (184, 189), (193, 198)]
        header_template_fpath = pkg_resources.resource_filename('htof', 'data/hip2_recalibrated_header.txt')
        copy(header_template_fpath, path)  # copy the template file to the output path.
        # populate the header lines.
        f = open(path, 'r')
        lines = f.readlines()
        for idx, hline, ks, bs, line_length in zip([6, 8, 10], ['first', 'second', 'third'],
                                                   [ks1, ks2, ks3], [bs1, bs2, bs3], [48, 34, 201]):
            matter = ["#"] + [" "] * (line_length - 1)  # strings are immutable in python. This is a gross solution to get past that.
            for key, byte_start_stop in zip(ks, bs):
                minn, maxx = min(byte_start_stop)-1, max(byte_start_stop)-1  # -1 because python indexes from 0.
                final_len = 1 + maxx-minn
                value = str(self.recalibrated_header[hline][key])
                matter[minn: maxx + 1] = list(value[:final_len].ljust(final_len, " "))
            matter = "".join(matter) # convert to a string instead of a list of characters.
            lines[idx] = matter + '\n'
            lines[idx] = lines[idx].replace('nan', '---')
        lines[-1] += '\n'
        # populate the IAD entries
        iad_bytes = [(3, 6), (8, 14), (16, 22), (24, 30), (32, 38), (42, 46), (49, 53)]
        significance = [None, 4, 4, 4, 4, 2, 2]  # None means values in that column will not be rounded.
        #
        line_length = 56
        for line in self.recalibrated_data.to_numpy():
            matter = [" "] * line_length  # strings are immutable in python. This is a gross solution to get past that.
            for col, value, byte_start_stop, sig in zip(np.arange(7), line, iad_bytes, significance):
                minn, maxx = min(byte_start_stop) - 1, max(byte_start_stop) - 1  # -1 because python indexes from 0.
                final_len = 1 + maxx - minn
                if col != 0:
                    # left justify every column except the orbit column. Round floats also.
                    value = value if sig is None else np.round(value, sig)
                    value = str(value) if value < 0 else ' ' + str(value)  # left pad with a space if it is positive.
                    matter[minn: maxx + 1] = list(value[:final_len].ljust(final_len, " "))
                else:
                    # right justify the orbit column
                    value = str(int(value))
                    matter[minn: maxx + 1] = list(value[:final_len].rjust(final_len, " "))
            matter = "".join(matter) + '\n'  # convert to a string instead of a list of characters.
            lines.append(matter)
        # write it out to file.
        f = open(path, 'w')
        f.writelines(lines)
        f.close()
        return None


class Hipparcos2ParserFactory:
    """
    A factory method for HipparcosRereductionDVDBook and HipparcosRereductionJavaTool. It detects
    which format the IAD is in, then chooses and returns the appropriate parser.
    """

    @staticmethod
    def get_appropriate_parser(filepath):
        if filepath == '':
            return HipparcosRereductionJavaTool
        datatype = get_datatype(filepath)
        if datatype == 'hip2dvd':
            return HipparcosRereductionDVDBook
        if datatype == 'hip2javatool':
            return HipparcosRereductionJavaTool

    @classmethod
    def parse_and_instantiate(cls, star_id: str, intermediate_data_directory: str, **kwargs):
        filepath, msg = DataParser.get_intermediate_data_file_path(star_id, intermediate_data_directory)
        CorrectDataParser = cls.get_appropriate_parser(filepath)
        return CorrectDataParser.parse_and_instantiate(star_id, intermediate_data_directory, **kwargs)


def to_ra_dec_basis(value, scan_angle):
    """
    Convert values (e.g. residuals) along the direction of the scan to the same value in RA and Dec. I.e. assume the value
    has a zero across_scan component and all of the value is in the along-scan direction, then convert that value to a vector in
    RA and Dec.
    these maths are just from https://en.wikipedia.org/wiki/Rotation_of_axes  .
    """
    dec_value, ra_value = value * np.cos(scan_angle), value * np.sin(scan_angle)
    return ra_value, dec_value


def to_along_scan_basis(ra_value, dec_value, scan_angle):
    """
    Convert values (e.g. residuals) in RA and Dec to the same value along the direction of the scan. I.e convert from RA, DEC to
    Along-scan , Across-scan basis, then keep only the along-scan component.
    these maths are just from https://en.wikipedia.org/wiki/Rotation_of_axes  .

    note that the inputs ra_value and dec_value may have nonzero across scan components. This function will zero
    out the cross-scan component. to_ra_dec_basis and to_along_scan_basis() are ONLY inverse transforms for data that
    has components solely in the across-scan direction.
    """
    along_scan_value = dec_value * np.cos(scan_angle) + ra_value * np.sin(scan_angle)
    return along_scan_value


def get_datatype(filepath):
    with open(filepath) as f:
        lines = f.readlines()
    if 'This file contains residual records' in lines[0]:
        datatype = 'hip2javatool'
    else:
        datatype = 'hip2dvd'
    return datatype

